substrate/frame/scheduler/src/lib.rs

```diff
@@ -1124,11 +1258,17 @@ impl<T: Config> Pallet<T> {
 			},
 			Err(()) => Err((Overweight, Some(task))),
```

Comment:
> in this and the Err arm above you abandon the retry entry if there is any

The `Err` arms above deal with completely different problems:
- `Err(Unavailable)` is a task which may never be able to run and it may never even be attempted again
- `Err(Overweight)` is a task which couldn't run because the scheduler ran out of weight, but it will be run the next time the scheduler services agendas; the scheduler runs agendas from previous blocks if they didn't execute completely, in order of block number from earliest up until present

This means that an overweight task is not abandoned and will be run again soon, but also that there is absolutely nothing that we can do about it. There is no fallback for the current block if we run out of weight. This is the scheduler's behavior before this PR and I believe the retry mechanism has nothing to do with this.

> I would move the retry handling one level above, for this you just need to add for Ok varian an info if it failed or not.

Most of the logic is common for failed or successful tasks, there would be a lot of copied code. But if you just want to refactor this, I could move the common logic out of the `match`.

> or at least take it before `execute_dispatch`

Tasks are scheduled for retries if they actually run but fail in their execution. By definition, there is no way to handle this before running `execute_dispatch`.

Reply:
I thought that the task is removed if `Unavailable`, but now I see that we keep it in the storage. This is why I proposed to remove the retry as well. And also I see now that `Overweight` does not change it's address.